arch/arm/core/aarch64/switch.S

```diff
@@ -23,34 +23,23 @@ GDATA(_kernel)
 GDATA(_k_neg_eagain)
 
 /**
- * @brief PendSV exception handler, handling context switches
+ * @brief Routine to handle context switches
  *
- * The PendSV exception is the only execution context in the system that can
- * perform context switching. When an execution context finds out it has to
- * switch contexts, it pends the PendSV exception.
- *
- * When PendSV is pended, the decision that a context switch must happen has
- * already been taken. In other words, when z_arm64_pendsv() runs, we *know* we
- * have to swap *something*.
- *
- * For Cortex-A, PendSV exception is not supported by the architecture and this
- * function is directly called either by _isr_wrapper() in case of preemption,
- * or z_arm64_svc() in case of cooperative switching.
+ * This function is directly called either by _isr_wrapper() in case of
+ * preemption, or z_arm64_svc() in case of cooperative switching.
  */
-GTEXT(z_arm64_pendsv)
-SECTION_FUNC(TEXT, z_arm64_pendsv)
-#ifdef CONFIG_TRACING
-	stp	xzr, x30, [sp, #-16]!
-	bl	sys_trace_thread_switched_in
-	ldp	xzr, x30, [sp], #16
-#endif
-	/* load _kernel into x1 and current k_thread into x2 */
-	ldr	x1, =_kernel
-	ldr	x2, [x1, #_kernel_offset_to_current]
+
+GTEXT(z_arm64_context_switch)
+SECTION_FUNC(TEXT, z_arm64_context_switch)
+	/*
+	 * dummy_thread->switch_handle is 0 when setup by z_cstart()
+	 */
+	cmp	x17, #0
+	beq	null_switch
```

Comment:
I wonder if there is a better way of handling this (i.e. something that does not require a `NULL` check whenever a context switch occurs)?

Reply:
@andyross might have some advice here

Reply:
The dummy thread should never be switched "to", so its switch handle isn't ever supposed to be inspected.  This sounds like something is working backwards here.   The idea behind arch_switch is that you don't muck with the scheduler internals.  The kernel decides for you what switch handle you should be switching to, and asks that you store the old one through the pointer in the second argument.